src/storage/src/render/persist_sink.rs

```diff
@@ -647,126 +656,126 @@ where
-
-                                    source_statistics.inc_updates_staged_by(1);
-
-                                    // Note that we assume `diff` is either +1 or -1 here, being anything
-                                    // else is a logic bug we can't handle at the metric layer. We also
-                                    // assume this addition doesn't overflow.
-                                    match (is_value, diff.is_positive()) {
-                                        (true, true) => {
-                                            builder.metrics.inserts += diff.unsigned_abs()
-                                        },
-                                        (true, false) => {
-                                            builder.metrics.retractions += diff.unsigned_abs()
-                                        },
-                                        (false, true) => {
-                                            builder.metrics.error_inserts += diff.unsigned_abs()
-                                        },
-                                        (false, false) => {
-                                            builder.metrics.error_retractions += diff.unsigned_abs()
-                                        },
-                                    }
+                            match in_flight_batches.entry(description) {
+                                std::collections::hash_map::Entry::Vacant(v) => {
+                                    // This _should_ be `.retain`, but rust
+                                    // currently thinks we can't use `cap`
+                                    // as an owned value when using the
+                                    // match guard `Some(event)`
+                                    v.insert(cap.delayed(cap.time()));
+                                }
+                                std::collections::hash_map::Entry::Occupied(o) => {
+                                    let (description, _) = o.remove_entry();
+                                    panic!(
+                                        "write_batches: sink {} got more than one \
+                                            batch for description {:?}, in-flight: {:?}",
+                                        collection_id, description, in_flight_batches
+                                    );
                                 }
                             }
+                        }
+                    }
+                    Event::Progress(frontier) => {
+                        batch_descriptions_frontier = frontier;
+                    }
+                }
+            }
 
-                            continue;
+            let ready_events = std::iter::from_fn(|| desired_input.next_sync()).collect_vec();
+
+            // We know start the async work for the input we received. Until we finish the dataflow
+            // should be marked as busy.
+            let permit = busy_signal.acquire().await;
```

Comment:
It's possible for `ready_events` to be empty here, right? (If the `select!` had moved forward because of the `descriptions_input` future resolving).

In that case we would only need to acquire the semaphore if there were also progress changes, so a slight optimization would be to avoid acquiring this unless absolutely needed

Reply:
You're right, it can be empty. Doing it conditionally would require doing something like this:

```rust
let permit = if ready_events.is_empty() {
    None
} else {
    Some(busy_signal.acquire.await)
};
```

Which.. is doable but also a bit verbose. Given that acquiring the semaphore is very cheap compared to everything else we're doing here and that there can't be contention do you mind if we optimize for readability and unconditionally acquire the semaphore?

Reply:
yeah totally fine - more of a suggestion if we think it's expensive to acquire the semaphore on every loop